File: main/dialogs.js
// ============================================================
// 主进程文件/对话框服务：原生选择器、目录扫描与二进制读写
// ============================================================
'use strict';

function registerDialogsIpc({ ipcMain, dialog, path, fs, app }) {
  // 原生对话框
  ipcMain.handle('dialog:pickAudio', async () => {
    const r = await dialog.showOpenDialog({
      properties: ['openFile'],
      filters: [{ name: '音频 / 视频', extensions: ['mp3', 'wav', 'flac', 'm4a', 'aac', 'ogg', 'opus', 'wma', 'mp4', 'mkv', 'avi', 'mov', 'webm'] }],
    });
    return r.canceled ? null : r.filePaths[0];
  });
  // 批量转录：一次选择多个音频 / 视频
  ipcMain.handle('dialog:pickAudioFiles', async () => {
    const r = await dialog.showOpenDialog({
      properties: ['openFile', 'multiSelections'],
      filters: [{ name: '音频 / 视频', extensions: ['mp3', 'wav', 'flac', 'm4a', 'aac', 'ogg', 'opus', 'wma', 'mp4', 'mkv', 'avi', 'mov', 'webm'] }],
    });
    return r.canceled ? [] : r.filePaths;
  });
  // 批量转录：递归列出文件夹内音频/视频（上限 2000 / 8 层）
  ipcMain.handle('dir:listAudioFiles', async (_e, dir) => {
    try {
      if (!dir || !fs.existsSync(dir)) return [];
      const exts = new Set(['.mp3', '.wav', '.flac', '.m4a', '.aac', '.ogg', '.oga', '.opus', '.wma', '.mp4', '.mkv', '.avi', '.mov', '.webm', '.aiff', '.aif', '.au', '.snd', '.caf', '.m4v', '.m4s', '.ts', '.mpg', '.mpeg', '.flv', '.3gp', '.amr', '.mka']);
      const out = [];
      const walk = (d, depth) => {
        if (out.length >= 2000 || depth > 8) return;
        let items;
        try { items = fs.readdirSync(d, { withFileTypes: true }); } catch { return; }
        for (const it of items) {
          if (out.length >= 2000) return;
          try {
            if (it.isDirectory()) walk(path.join(d, it.name), depth + 1);
            else if (it.isFile() && exts.has(path.extname(it.name).toLowerCase())) out.push(path.join(d, it.name));
          } catch {}
        }
      };
      walk(dir, 0);
      out.sort((a, b) => a.localeCompare(b, 'zh'));
      return out;
    } catch (e) { return []; }
  });
  ipcMain.handle('dialog:pickImage', async () => {
    const r = await dialog.showOpenDialog({
      properties: ['openFile'],
      filters: [{ name: '图片', extensions: ['png', 'jpg', 'jpeg', 'bmp', 'webp', 'gif'] }],
    });
    return r.canceled ? null : r.filePaths[0];
  });
  ipcMain.handle('dialog:pickFile', async (_e, opts) => {
    const r = await dialog.showOpenDialog({
      properties: ['openFile'],
      filters: (opts && opts.filters) || [{ name: '文件', extensions: ['*'] }],
    });
    return r.canceled ? null : r.filePaths[0];
  });
  ipcMain.handle('dialog:pickDirectory', async () => {
    const r = await dialog.showOpenDialog({ properties: ['openDirectory'] });
    return r.canceled ? null : r.filePaths[0];
  });
  // 模型本地压缩包导入（.zip / .7z / .tar）
  ipcMain.handle('dialog:pickModelArchive', async () => {
    const r = await dialog.showOpenDialog({
      properties: ['openFile'],
      filters: [{ name: '模型压缩包', extensions: ['zip', '7z', 'tar', 'gz', 'tgz', 'tar.gz', 'txz', '001', '002', '003', 'z01', 'z02'] }],
    });
    return r.canceled ? null : r.filePaths[0];
  });
  // 内置 SoundFont 列表（随应用分发，用户直接选择，无需加载外部文件）
  ipcMain.handle('soundfont:list', async () => {
    try {
      const dir = path.join(__dirname, '..', 'renderer', 'vendor', 'soundfonts');
      if (!fs.existsSync(dir)) return [];
      return fs.readdirSync(dir)
        .filter(f => /\.(sf2|sf3)$/i.test(f))
        .map(f => ({ id: f, name: f.replace(/\.[^.]+$/, ''), path: path.join(dir, f) }));
    } catch (e) { return []; }
  });
  // MusicXML 导入
  ipcMain.handle('dialog:pickMusicXML', async () => {
    const r = await dialog.showOpenDialog({
      properties: ['openFile'],
      filters: [{ name: 'MusicXML', extensions: ['xml', 'musicxml', 'mxl'] }],
    });
    return r.canceled ? null : r.filePaths[0];
  });

  // 歌单“导入文件夹”：递归列出目录下的 MIDI 文件（上限 2000 / 8 层，避免误选整盘卡死）
  ipcMain.handle('dir:listMidiFiles', async (_e, dir) => {
    try {
      if (!dir || !fs.existsSync(dir)) return [];
      const exts = new Set(['.mid', '.midi', '.kar', '.rmi']);
      const out = [];
      const walk = (d, depth) => {
        if (out.length >= 2000 || depth > 8) return;
        let items;
        try { items = fs.readdirSync(d, { withFileTypes: true }); } catch { return; }
        for (const it of items) {
          if (out.length >= 2000) return;
          try {
            if (it.isDirectory()) walk(path.join(d, it.name), depth + 1);
            else if (it.isFile() && exts.has(path.extname(it.name).toLowerCase())) out.push(path.join(d, it.name));
          } catch {}
        }
      };
      walk(dir, 0);
      out.sort((a, b) => a.localeCompare(b, 'zh'));
      return out;
    } catch (e) { return []; }
  });
  // 读取本地文件（转录结果回载）——异步 + 大小上限，避免阻塞主进程
  ipcMain.handle('file:readBinary', async (_e, p) => {
    try {
      const st = await fs.promises.stat(p);
      if (!st.isFile() || st.size > 64 * 1024 * 1024) return null;
      const buf = await fs.promises.readFile(p);
      return buf.buffer.slice(buf.byteOffset, buf.byteOffset + buf.byteLength);
    } catch { return null; }
  });
  // 读取 SoundFont（SF2/SF3 通常数十 MB 到数百 MB，单独放宽上限）
  ipcMain.handle('file:readSoundFont', async (_e, p) => {
    try {
      const st = await fs.promises.stat(p);
      if (!st.isFile() || st.size > 512 * 1024 * 1024) return null;
      const buf = await fs.promises.readFile(p);
      return buf.buffer.slice(buf.byteOffset, buf.byteOffset + buf.byteLength);
    } catch { return null; }
  });

  // 保存二进制（WAV / MIDI 导出）：原生保存对话框 → fs 写盘。
  // 绕开 Electron 33 失效的下载子系统：<a download> 经 will-download 的
  // preventDefault+setSavePath+resume 后下载会被 0 字节取消（blob/data/http 均如此）。
  ipcMain.handle('file:saveBinary', async (_e, opts) => {
    try {
      const name = (opts && opts.name) || 'download.bin';
      const ext = path.extname(name).toLowerCase() || '';
      const filters = ext === '.wav'
        ? [{ name: 'WAV 音频', extensions: ['wav'] }]
        : ext === '.mid' || ext === '.midi'
          ? [{ name: 'MIDI 音频', extensions: ['mid', 'midi'] }]
          : [{ name: '文件', extensions: ['*'] }];
      const r = await dialog.showSaveDialog({
        title: '保存文件',
        defaultPath: path.join(app.getPath('downloads'), name),
        filters,
      });
      if (r.canceled || !r.filePath) return { ok: false, canceled: true };
      const buf = (opts && opts.data != null) ? Buffer.from(opts.data) : Buffer.alloc(0);
      await fs.promises.writeFile(r.filePath, buf);
      return { ok: true, path: r.filePath };
    } catch (e) { return { ok: false, canceled: false, error: String((e && e.message) || e) }; }
  });
}

module.exports = { registerDialogsIpc };

File: src-tauri/src/dialogs.rs
// 文件对话框 / 目录扫描 / 二进制读写 —— 移植自 main/dialogs.js
use serde::Deserialize;
use serde_json::{json, Value};
use std::fs;
use std::path::Path;
use tauri::Manager;
use tauri_plugin_dialog::DialogExt;

const AUDIO_EXTS: &[&str] = &[
    "mp3", "wav", "flac", "m4a", "aac", "ogg", "oga", "opus", "wma", "mp4", "mkv", "avi", "mov",
    "webm", "aiff", "aif", "au", "snd", "caf", "m4v", "m4s", "ts", "mpg", "mpeg", "flv", "3gp",
    "amr", "mka",
];
const MIDI_EXTS: &[&str] = &["mid", "midi", "kar", "rmi"];

fn has_ext(p: &Path, exts: &[&str]) -> bool {
    match p.extension().and_then(|e| e.to_str()) {
        Some(e) => exts.contains(&e.to_lowercase().as_str()),
        None => false,
    }
}

fn walk(dir: &Path, exts: &[&str], out: &mut Vec<String>, depth: usize) {
    if out.len() >= 2000 || depth > 8 {
        return;
    }
    let items = match fs::read_dir(dir) {
        Ok(i) => i,
        Err(_) => return,
    };
    for it in items.flatten() {
        if out.len() >= 2000 {
            return;
        }
        let p = it.path();
        if p.is_dir() {
            walk(&p, exts, out, depth + 1);
        } else if p.is_file() && has_ext(&p, exts) {
            out.push(p.to_string_lossy().to_string());
        }
    }
}

// ---- 对话框命令（使用 tauri-plugin-dialog，父窗口为当前主窗口）----

#[tauri::command]
pub async fn pick_audio(app: tauri::AppHandle, window: tauri::WebviewWindow) -> Option<String> {
    let file = window
        .dialog()
        .file()
        .add_filter("音频 / 视频", AUDIO_EXTS)
        .blocking_pick_file();
    file.and_then(|f| f.into_path().ok()).map(|p| p.to_string_lossy().to_string())
}

#[tauri::command]
pub async fn pick_audio_files(
    app: tauri::AppHandle,
    window: tauri::WebviewWindow,
) -> Vec<String> {
    let files = window
        .dialog()
        .file()
        .add_filter("音频 / 视频", AUDIO_EXTS)
        .blocking_pick_files();
    files
        .into_iter()
        .flatten()
        .filter_map(|f| f.into_path().ok())
        .map(|p| p.to_string_lossy().to_string())
        .collect()
}

#[tauri::command]
pub async fn list_audio_files(dir: String) -> Vec<String> {
    let mut out = Vec::new();
    walk(Path::new(&dir), AUDIO_EXTS, &mut out, 0);
    out.sort();
    out
}

#[tauri::command]
pub async fn pick_image(app: tauri::AppHandle, window: tauri::WebviewWindow) -> Option<String> {
    let file = window
        .dialog()
        .file()
        .add_filter("图片", &["png", "jpg", "jpeg", "bmp", "webp", "gif"])
        .blocking_pick_file();
    file.and_then(|f| f.into_path().ok()).map(|p| p.to_string_lossy().to_string())
}

#[derive(Deserialize)]
pub struct PickFileOpts {
    filters: Option<Vec<FileFilterOpts>>,
}
#[derive(Deserialize)]
pub struct FileFilterOpts {
    name: Option<String>,
    extensions: Option<Vec<String>>,
}

#[tauri::command]
pub async fn pick_file(
    app: tauri::AppHandle,
    window: tauri::WebviewWindow,
    opts: Option<PickFileOpts>,
) -> Option<String> {
    let mut dlg = window.dialog().file();
    if let Some(o) = opts {
        if let Some(fs) = o.filters {
            for f in fs {
                let exts: Vec<String> = f
                    .extensions
                    .unwrap_or_default()
                    .iter()
                    .map(|s| s.clone())
                    .collect();
                let exts_ref: Vec<&str> = exts.iter().map(|s| s.as_str()).collect();
                dlg = dlg.add_filter(f.name.unwrap_or_else(|| "文件".into()), &exts_ref);
            }
        }
    }
    let file = dlg.blocking_pick_file();
    file.and_then(|f| f.into_path().ok()).map(|p| p.to_string_lossy().to_string())
}

#[tauri::command]
pub async fn pick_directory(
    app: tauri::AppHandle,
    window: tauri::WebviewWindow,
) -> Option<String> {
    let dir = window.dialog().file().blocking_pick_folder();
    dir.and_then(|f| f.into_path().ok()).map(|p| p.to_string_lossy().to_string())
}

#[tauri::command]
pub async fn soundfont_list(app: tauri::AppHandle) -> Vec<Value> {
    // 打包后 soundfonts 位于 resources/renderer/vendor/soundfonts
    let mut roots: Vec<std::path::PathBuf> = Vec::new();
    if let Ok(res) = app.path().resource_dir() {
        roots.push(res.join("renderer").join("vendor").join("soundfonts"));
    }
    let mut out = Vec::new();
    for dir in roots {
        if let Ok(entries) = fs::read_dir(&dir) {
            for e in entries.flatten() {
                let p = e.path();
                let name = p
                    .file_name()
                    .map(|n| n.to_string_lossy().to_string())
                    .unwrap_or_default();
                if name.ends_with(".sf2") || name.ends_with(".sf3") {
                    let id = name
                        .trim_end_matches(".sf2")
                        .trim_end_matches(".sf3")
                        .to_string();
                    out.push(json!({ "id": name, "name": id, "path": p.to_string_lossy() }));
                }
            }
        }
    }
    out
}

#[tauri::command]
pub async fn pick_music_xml(
    app: tauri::AppHandle,
    window: tauri::WebviewWindow,
) -> Option<String> {
    let file = window
        .dialog()
        .file()
        .add_filter("MusicXML", &["xml", "musicxml", "mxl"])
        .blocking_pick_file();
    file.and_then(|f| f.into_path().ok()).map(|p| p.to_string_lossy().to_string())
}

#[tauri::command]
pub async fn list_midi_files(dir: String) -> Vec<String> {
    let mut out = Vec::new();
    walk(Path::new(&dir), MIDI_EXTS, &mut out, 0);
    out.sort();
    out
}

// 读取二进制（上限 64MB）
#[tauri::command]
pub async fn read_binary(p: String) -> Option<Vec<u8>> {
    let meta = fs::metadata(&p).ok()?;
    if !meta.is_file() || meta.len() > 64 * 1024 * 1024 {
        return None;
    }
    fs::read(&p).ok()
}

// 读取 SoundFont（上限 512MB）
#[tauri::command]
pub async fn read_soundfont(p: String) -> Option<Vec<u8>> {
    let meta = fs::metadata(&p).ok()?;
    if !meta.is_file() || meta.len() > 512 * 1024 * 1024 {
        return None;
    }
    fs::read(&p).ok()
}

#[derive(Deserialize)]
pub struct SaveBinaryOpts {
    name: Option<String>,
    data: Option<Vec<u8>>,
}

// 保存二进制（原生保存对话框 → 写盘）
#[tauri::command]
pub async fn save_binary(
    app: tauri::AppHandle,
    window: tauri::WebviewWindow,
    opts: Option<SaveBinaryOpts>,
) -> Value {
    let name = opts.as_ref().and_then(|o| o.name.clone()).unwrap_or_else(|| "download.bin".into());
    let ext = Path::new(&name)
        .extension()
        .and_then(|e| e.to_str())
        .unwrap_or("")
        .to_lowercase();
    let filters: &[&str] = if ext == "wav" {
        &["wav"]
    } else if ext == "mid" || ext == "midi" {
        &["mid", "midi"]
    } else {
        &[]
    };
    let mut dlg = window.dialog().file();
    if !filters.is_empty() {
        dlg = dlg.add_filter("保存文件", filters);
    }
    let saved = dlg.blocking_save_file();
    match saved {
        Some(file_path) => {
            let data = opts.and_then(|o| o.data).unwrap_or_default();
            if let Ok(p) = file_path.into_path() {
                if fs::write(&p, data).is_ok() {
                    return json!({ "ok": true, "path": p.to_string_lossy() });
                }
                return json!({ "ok": false, "error": "写入失败" });
            }
            json!({ "ok": false, "canceled": true })
        }
        None => json!({ "ok": false, "canceled": true }),
    }
}
